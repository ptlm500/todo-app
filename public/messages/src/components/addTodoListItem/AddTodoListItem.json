[
  {
    "id": "todo.addTodo.namePlaceholder",
    "description": "Placeholder message for the add todo name field",
    "defaultMessage": "Enter a name"
  },
  {
    "id": "todo.addTodo.descriptionPlaceholder",
    "description": "Placeholder message for the add todo description field",
    "defaultMessage": "Enter a description"
  },
  {
    "id": "todo.addTodo.cancel",
    "description": "Add todo cancel button text",
    "defaultMessage": "Cancel"
  },
  {
    "id": "todo.addTodo.add",
    "description": "Add todo add button text",
    "defaultMessage": "Add"
  }
]