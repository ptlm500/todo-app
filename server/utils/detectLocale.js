const acceptLanguage = require('accept-language');
const config = require('../config.json');

function detectLocale(req) {
  acceptLanguage.languages(config.ACCEPTED_LANGUAGES);

  const cookieLocale = req.cookies.locale;

  return acceptLanguage.get(cookieLocale || req.headers['accept-language']) ||
    config.DEFAULT_LOCALE;
}

module.exports = detectLocale;
