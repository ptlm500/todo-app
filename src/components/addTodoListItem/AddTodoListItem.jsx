import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTodo } from '../../state/actions/actions';
import TextareaAutosize from 'react-autosize-textarea';
import { v4 as uuid } from 'uuid';
import {
  PrimaryButton,
  SecondaryButton
} from '../button';

import './addTodoListItem.scss';

function generateTodo(name, description) {
  return {
    id: uuid(),
    name,
    description,
    creationDate: '' + Date.now()
  };
}

export function AddTodoListItem(props) {
  const {
    onAddTodo
  } = props;

  const [addingItem, setAddingItem] = useState(false);
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');

  function submitTodo() {
    onAddTodo(generateTodo(name, description));
    setAddingItem(false);
  }

  function cancel() {
    setName('');
    setDescription('');
    setAddingItem(false);
  }

  return (
    !addingItem ? (
      <PrimaryButton onClick={() => setAddingItem(true)}>New todo</PrimaryButton>
    ) : (
      <div className='todo-list__new-item'>
        <input
          placeholder={'Enter a name'}
          value={name}
          onChange={e => setName(e.target.value)}
        />
        <TextareaAutosize
          placeholder={'Enter a description'}
          value={description}
          onChange={e => setDescription(e.target.value)}
        />
        <div className='todo-list__new-item-actions'>
          <SecondaryButton onClick={cancel}>Cancel</SecondaryButton>
          <PrimaryButton onClick={submitTodo}>Add</PrimaryButton>
        </div>
      </div>
    )
  );
}

AddTodoListItem.propTypes = {
  onAddTodo: PropTypes.func.isRequired
};

const mapDispatchToProps = dispatch => {
  return {
    onAddTodo: todo => dispatch(addTodo(todo))
  };
};

export default connect(
  null,
  mapDispatchToProps
)(AddTodoListItem);
