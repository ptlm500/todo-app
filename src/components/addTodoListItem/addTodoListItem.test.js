import React from 'react';
import { act } from 'react-test-renderer';
import { renderWithIntl } from '../../../test/utils';
import { AddTodoListItem } from './AddTodoListItem';
import 'jest-styled-components';

// Mocking transition group https://github.com/reactjs/react-transition-group/issues/208
jest.mock('react-transition-group', () => {
  /* eslint-disable react/prop-types */
  return {
    CSSTransition: function CSSTransition(props) {
      return (
        <div
          id="CSSTransition"
          {...props}
        >
          {props.children}
        </div>
      );
    }
  };
  /* eslint-enable react/prop-types */
});

jest.mock('uuid', () => {
  return {
    v4: jest.fn(() => 'test-uuid')
  };
});

const defaultProps = {
  onAddTodo: jest.fn()
};

const realNow = Date.now;

beforeAll(() => {
  global.Date.now = jest.fn(() => 1582069219000);
});

beforeEach(() => {
  jest.clearAllMocks();
});

afterAll(() => {
  global.Date.now = realNow;
});

describe('AddTodoListItem', () => {
  it('matches the snapshot', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    expect(tree).toMatchSnapshot();
  });

  it('hides the addItem button and displays the form when the addItem button is clicked', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    const addItemButton = findAddItemButton(tree);
    const transitionGroup = findTransitionGroup(tree);

    act(() => {
      addItemButton.props.onClick();
    });

    expect(transitionGroup.props.appear).toBeTruthy();
  });

  it('updates the name in state on NameInput change', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    const nameInput = findNameInput(tree);

    const testInput = 'test input';
    act(() => {
      nameInput.props.onChange({target: {value: testInput}});
    });

    expect(nameInput.props.value).toBe(testInput);
  });

  it('updates the name in state on DescriptionInput change', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    const descriptionInput = findDescriptionInput(tree);

    const testInput = 'test input';
    act(() => {
      descriptionInput.props.onChange({target: {value: testInput}});
    });

    expect(descriptionInput.props.value).toBe(testInput);
  });

  it('hides the form and clears name and description when the cancel button is clicked', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    const addItemButton = findAddItemButton(tree);
    const cancelButton = findCancelButton(tree);
    const transitionGroup = findTransitionGroup(tree);
    const nameInput = findNameInput(tree);
    const descriptionInput = findDescriptionInput(tree);

    act(() => {
      addItemButton.props.onClick();
      nameInput.props.onChange({target: {value: 'test'}});
      descriptionInput.props.onChange({target: {value: 'test'}});
    });

    act(() => {
      cancelButton.props.onClick();
    });

    // Test form hidden
    expect(transitionGroup.props.in).toBeFalsy();
    // Test state is cleared
    expect(nameInput.props.value).toBe('');
    expect(descriptionInput.props.value).toBe('');
  });

  it('calls onAddTodo, hides the form and clears name and description when the add button is clicked', () => {
    let tree;
    act(() => {
      tree = renderWithIntl(<AddTodoListItem {...defaultProps} />);
    });

    const addItemButton = findAddItemButton(tree);
    const addButton = findAddButton(tree);
    const transitionGroup = findTransitionGroup(tree);
    const nameInput = findNameInput(tree);
    const descriptionInput = findDescriptionInput(tree);

    const testName = 'test name';
    const testDescription = 'test description';

    act(() => {
      addItemButton.props.onClick();
      nameInput.props.onChange({target: {value: testName}});
      descriptionInput.props.onChange({target: {value: testDescription}});
    });

    act(() => {
      addButton.props.onClick();
    });

    // Test onAddTodo call
    expect(defaultProps.onAddTodo).toHaveBeenCalledWith({
      id: 'test-uuid',
      name: testName,
      description: testDescription,
      creationDate: '1582069219000'
    });
    // Test form hidden
    expect(transitionGroup.props.in).toBeFalsy();
    // Test state is cleared
    expect(nameInput.props.value).toBe('');
    expect(descriptionInput.props.value).toBe('');
  });
});

function findAddItemButton(tree) {
  return tree.root.findByProps({id: 'add-item'});
}

function findTransitionGroup(tree) {
  return tree.root.findByProps({id: 'CSSTransition'});
}

function findNameInput(tree) {
  return tree.root.findByProps({id: 'name-input'});
}

function findDescriptionInput(tree) {
  return tree.root.findByProps({id: 'description-input'});
}

function findCancelButton(tree) {
  return tree.root.findByProps({id: 'cancel'});
}

function findAddButton(tree) {
  return tree.root.findByProps({id: 'add'});
}
