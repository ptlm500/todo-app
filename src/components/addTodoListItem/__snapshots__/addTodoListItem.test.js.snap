// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodoListItem matches the snapshot 1`] = `
Array [
  .c0 {
  background: none;
  border: none;
  width: 3rem;
  height: 3rem;
  padding: 0;
  cursor: pointer;
  -webkit-transition: 0.3s;
  transition: 0.3s;
}

.c0 > svg {
  width: 100%;
  height: 100%;
}

.c0:hover {
  -webkit-transform: scale(1.2) rotate(1turn);
  -ms-transform: scale(1.2) rotate(1turn);
  transform: scale(1.2) rotate(1turn);
}

@media (max-width:599px) {

}

@media (max-width:240px) {

}

@media (min-width:640px) {

}

@media (max-width:480px) {

}

@media (min-width:640px) {

}

<button
    className="c0"
    id="add-item"
    onClick={[Function]}
    size="large"
  >
    <svg
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        New todo
      </title>
      <linearGradient
        gradientTransform="matrix(1 0 0 -1 0 -12310)"
        gradientUnits="userSpaceOnUse"
        id="a"
        x1="0"
        x2="512"
        y1="-12566"
        y2="-12566"
      >
        <stop
          offset="0"
          stopColor="#303F9F"
        />
        <stop
          offset="1"
          stopColor="#D32F2F"
        />
      </linearGradient>
      <path
        d="m512 256c0 141.386719-114.613281 256-256 256s-256-114.613281-256-256 114.613281-256 256-256 256 114.613281 256 256zm0 0"
        fill="url(#a)"
      />
      <g
        fill="#fff"
      >
        <path
          d="m223.085938 411h-48.183594c-15.558594 0-28.21875-12.660156-28.21875-28.21875v-253.5625c0-15.558594 12.660156-28.21875 28.21875-28.21875h177.125c15.5625 0 28.21875 12.660156 28.21875 28.21875v88.601562c0 8.285157 6.71875 15 15 15 8.285156 0 15-6.714843 15-15v-88.601562c0-32.101562-26.117188-58.21875-58.21875-58.21875h-177.121094c-32.105469 0-58.222656 26.117188-58.222656 58.21875v253.5625c0 32.101562 26.121094 58.21875 58.222656 58.21875h48.183594c8.28125 0 15-6.714844 15-15s-6.71875-15-15.003906-15zm0 0"
        />
        <path
          d="m337.621094 157.441406h-148.390625c-8.28125 0-15 6.714844-15 15s6.71875 15 15 15h148.390625c8.285156 0 15-6.714844 15-15s-6.714844-15-15-15zm0 0"
        />
        <path
          d="m352.621094 230.066406c0-8.28125-6.714844-15-15-15h-148.390625c-8.28125 0-15 6.71875-15 15 0 8.285156 6.71875 15 15 15h148.390625c8.285156 0 15-6.714844 15-15zm0 0"
        />
        <path
          d="m189.230469 272.695312c-8.285157 0-15 6.714844-15 15 0 8.285157 6.714843 15 15 15h90.125c8.28125 0 15-6.714843 15-15 0-8.285156-6.71875-15-15-15zm0 0"
        />
        <path
          d="m435.59375 279.765625c-17.082031-17.085937-44.878906-17.082031-61.949219-.011719l-79.097656 78.917969c-1.746094 1.746094-3.039063 3.894531-3.757813 6.261719l-17.222656 56.707031c-1.589844 5.234375-.207031 10.917969 3.613282 14.832031 2.863281 2.933594 6.75 4.527344 10.738281 4.527344 1.335937 0 2.679687-.179688 4.003906-.546875l58.148437-16.105469c2.492188-.691406 4.761719-2.015625 6.589844-3.839844l78.933594-78.78125c17.082031-17.082031 17.082031-44.878906 0-61.960937zm-21.210938 21.210937c5.382813 5.386719 5.382813 14.152344.007813 19.527344l-2.371094 2.367188-19.535156-19.535156 2.359375-2.359376c5.386719-5.382812 14.148438-5.382812 19.539062 0zm-76.121093 95.511719-28.066407 7.773438 8.207032-27.007813 52.84375-52.730468 19.539062 19.539062zm0 0"
        />
      </g>
    </svg>
  </button>,
  .c5 {
  background: #3F51B5;
  border-radius: 0.5rem;
  border: 2px solid #3F51B5;
  cursor: pointer;
  color: #FFFFFF;
  height: 2rem;
  padding: 0.25rem 1rem;
  -webkit-transition: 0.1s;
  transition: 0.1s;
  fill: #FFFFFF;
}

.c5 > svg {
  width: 1.5rem;
  width: 1.5rem;
  fill: currentColor;
}

.c5:hover:not([disabled]) {
  background: #303F9F;
  border: 2px solid #303F9F;
}

.c5:disabled {
  border: 2px solid #BDBDBD;
  background: #BDBDBD;
  color: #BDBDBD;
  fill: #3F51B5;
}

.c4 {
  background: #3F51B5;
  border-radius: 0.5rem;
  border: 2px solid #3F51B5;
  cursor: pointer;
  color: #FFFFFF;
  height: 2rem;
  padding: 0.25rem 1rem;
  -webkit-transition: 0.1s;
  transition: 0.1s;
  fill: #FFFFFF;
  background: #FFFFFF;
  border: 2px solid #3F51B5;
  color: #3F51B5;
  fill: #3F51B5;
}

.c4 > svg {
  width: 1.5rem;
  width: 1.5rem;
  fill: currentColor;
}

.c4:hover:not([disabled]) {
  background: #303F9F;
  border: 2px solid #303F9F;
}

.c4:disabled {
  border: 2px solid #BDBDBD;
  background: #BDBDBD;
  color: #BDBDBD;
  fill: #3F51B5;
}

.c4:hover:not([disabled]) {
  background: #3F51B5;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.c4:disabled {
  background: #FFFFFF;
  border: 2px solid #BDBDBD;
  cursor: auto;
}

.c0 {
  background: #C5CAE9;
  border-radius: 0.5rem;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-width: 500px;
  max-width: -webkit-fit-content;
  max-width: -moz-fit-content;
  max-width: fit-content;
}

.c0.card-transition-enter {
  opacity: 0.01;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c0.card-transition-enter-active {
  opacity: 1;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
  -webkit-transition: 200ms ease-out;
  transition: 200ms ease-out;
}

.c0.card-transition-appear {
  opacity: 0.01;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c0.card-transition-appear-active {
  opacity: 1;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
  -webkit-transition: 200ms ease-out;
  transition: 200ms ease-out;
}

.c0.card-transition-enter {
  opacity: 0.01;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c0.card-transition-enter-active {
  opacity: 1;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
  -webkit-transition: 200ms ease-out;
  transition: 200ms ease-out;
}

.c0.card-transition-appear {
  opacity: 0.01;
  -webkit-transform: scaleX(0) scaleY(0);
  -ms-transform: scaleX(0) scaleY(0);
  transform: scaleX(0) scaleY(0);
}

.c0.card-transition-appear-active {
  opacity: 1;
  -webkit-transform: scaleX(1) scaleY(1);
  -ms-transform: scaleX(1) scaleY(1);
  transform: scaleX(1) scaleY(1);
  -webkit-transition: 200ms ease-out;
  transition: 200ms ease-out;
}

.c1 {
  font-size: 5vw;
  font-weight: 600;
  margin-bottom: 1rem;
  max-width: 100%;
}

.c2 {
  background: #FFFFFF;
  color: #212121;
  font: inherit;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  border: 1px solid #BDBDBD;
  padding: 0px;
  width: 100%;
  max-width: 100%;
  font-size: 2.5vw;
  margin-bottom: 1rem;
  max-width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c3 > * {
  margin: 0 0.25rem;
}

.c3 > *:first-child {
  margin: 0 0.5rem 0 0;
}

.c3 > *:last-child {
  margin: 0 0 0 0.25rem;
}

@media (max-width:599px) {
  .c0 {
    max-width: 90%;
  }
}

@media (max-width:240px) {
  .c1 {
    font-size: 12px;
  }
}

@media (min-width:640px) {
  .c1 {
    font-size: 32px;
  }
}

@media (max-width:480px) {
  .c2 {
    font-size: 12px;
  }
}

@media (min-width:640px) {
  .c2 {
    font-size: 16px;
  }
}

<div
    appear={true}
    classNames="card-transition"
    exit={false}
    id="CSSTransition"
    in={false}
    timeout={200}
    unmountOnExit={true}
  >
    <div
      className="sc-AxirZ sc-AxiKw c0"
    >
      <input
        className="c1"
        id="name-input"
        onChange={[Function]}
        placeholder="Enter a name"
        value=""
      />
      <textarea
        async={false}
        className="sc-AxjAm c2"
        id="description-input"
        onChange={[Function]}
        placeholder="Enter a description"
        rows={1}
        value=""
      />
      <div
        className="c3"
      >
        <button
          className="c4"
          id="cancel"
          onClick={[Function]}
        >
          Cancel
        </button>
        <button
          className="c5"
          id="add"
          onClick={[Function]}
        >
          Add
        </button>
      </div>
    </div>
  </div>,
]
`;
